### PR TITLE
RooFit plotting enhancements

### PR DESCRIPTION
**Pull request by cburgard**
The RooFit "plotOn" plotting engine is very useful to visualize RooFit objects.
However, there is currently a strict separation between "Data-like" objects (which can only really be drawn with points) and "Function-like" objects (which can only really be drawn as lines and areas).
However, in techniques like unfolding, "data" is often times corrected data, and hence might actually be a function-type object in RooFit.
With this PR, functionality is added to RooAbsReal::plotOn that allows to plot functions "data-like", with data points including error bars when using the draw option "P", which was previously unusable.

### CODE REVIEW

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by cburgard** [comment]
I see that the CI is still failing through `clang-tidy`, but I cannot find any fault of my own here, seems to be related to a call to the virtual `TMatrixT::Clear` in the `TMatrixT` destructor...
Any opinions?

File: roofit/roofitcore/src/RooCurve.cxx [hunk 1]
```diff
@@ -118,13 +118,21 @@ RooCurve::RooCurve(const RooAbsReal &f, RooAbsRealLValue &x, Double_t xlo, Doubl
...
+    list<Double_t>* hint = f.plotSamplingHint(x,xlo,xhi) ;
+    addPoints(*funcPtr,xlo,xhi,xbins+1,prec,resolution,wmode,nEvalError,doEEVal,eeVal,hint);
+    if (_showProgress) {
+      ccoutP(Plotting) << endl ;
+    }
+    if (hint) {
+      delete hint ;
+    }
+  } else {
+    xbins = x.numBins();
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooCurve.cxx]: Is there a need to reset `xbins` to the new value here? Wouldn't a local variable that can be optimised out be sufficient?
Something like:
```
const size_t nBinsX = x.numBins();
<for loop until nBinsX>
```

File: roofit/roofitcore/src/RooAbsReal.cxx [hunk 2]
```diff
@@ -2277,25 +2283,42 @@ RooPlot* RooAbsReal::plotOn(RooPlot *frame, PlotOpt o) const
       // Append any suffixes imported from RooAbsPdf::plotOn
       curveName.Append(o.curveNameSuffix) ;
     }
-    curve->SetName(curveName.Data()) ;
+    
+    TString opt(o.drawOptions);
+    if(opt.Contains("P")){
+      RooHist *graph= new RooHist(*projection,*plotVar,1.,1.,frame->getNormVars(),o.errorFR);
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooAbsReal.cxx]: Isn't it advisable to disable evaluation errors similar to what's happening below? Or am I misunderstanding, and the PDF is not being evaluated? Or is it clear that there won't be such errors?

File: roofit/roofitcore/src/RooCurve.cxx (hunk 1 above)
Comment (cburgard) [reply roofit/roofitcore/src/RooCurve.cxx]: good suggestion, agreed.

File: roofit/roofitcore/src/RooAbsReal.cxx (hunk 2 above)
Comment (cburgard) [reply roofit/roofitcore/src/RooAbsReal.cxx]: you're right, this should be done.

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by hageboeck** [comment]
Hi Carsten,
it looks like the windows host gets stuck because it's trying to checkout roottest from your github. Could you fork it?

**Comment by cburgard** [comment]
done.
While playing with the code, I found that apparently my idea of how the `_nominalBinWidth` member is supposed to work was wrong - the normalization seems to be a bit off. I only get the correct behavior if I put `_nominalBinWidth=1`, but that seems wrong.
Any idea what the "right" thing to be doing here is?

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by hageboeck** [comment]
Hi Carsten,

The code seems ok, but the option is not documented. For which classes can you use it? Should it be added to the docs of RooAbsPdf::plotOn()?

**Comment by cburgard** [comment]
As far as documentation goes, RooFit doesn't really excessively document the draw options yet. Since it does just refer to TGraph, I naturally assumed that it would support all DrawOptions of TGraph both for RooAbsData and RooAbsReal, which turned out not to be the case, which is the whole reason for the MR.
Now at least RooAbsReal supports "P/E" instead of just "L/E", which is already a step forward, but RooAbsData still only supports "P/E" I think since I didn't mess with that. It might be useful to add this fact to the respective documentation lines of "DrawOption".
But since this was not documented so far at all, the old documentation doesn't become invalidated by this patch.
What might be useful though is to add the lines
MarkerStyle(Int_t style) | Select the ROOT marker style, default is 21
MarkerColor(Int_t color) | Select the ROOT marker color, default is black
MarkerSize(Double_t size) | Select the ROOT marker size
from the RooAbsData documentation table of plotOn also to RooAbsReal, since for the draw option "P", they are now supported there.

Regards,
Carsten

File: roofit/roofitcore/src/RooPlot.cxx [hunk 3]
```diff
@@ -266,7 +266,45 @@ RooPlot::RooPlot(const RooAbsRealLValue &var, Double_t xmin, Double_t xmax, Int_
   _normBinWidth = (xmax-xmin)/nbins ;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+//
+
+RooPlot* RooPlot::frame(const RooAbsRealLValue &var, Double_t xmin, Double_t xmax, Int_t nBins){
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooPlot.cxx]: Why this static function? Wouldn't it be better to call the constructor directly?
If the function is necessary, it should at least be documented for doxygen.

File: roofit/roofitcore/src/RooPlot.cxx [hunk 4]
```diff
@@ -266,7 +266,45 @@ RooPlot::RooPlot(const RooAbsRealLValue &var, Double_t xmin, Double_t xmax, Int_
...
+  return new RooPlot(var,xmin,xmax,nBins);
+}
+RooPlot* RooPlot::frameWithLabels(const RooAbsRealLValue &var){
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooPlot.cxx]: I would also advise to add doxygen comments, unless it's private/protected. Even if it is, it's nice to have them.

File: roofit/roofitcore/src/RooHist.cxx [hunk 5]
```diff
@@ -260,6 +261,65 @@ RooHist::RooHist(const RooHist& hist1, const RooHist& hist2, Double_t wgt1, Doub
 }
 
 
+RooHist::RooHist(const RooAbsReal &f, RooAbsRealLValue &x, Double_t xErrorFrac, Double_t scaleFactor, const RooArgSet *normVars, const RooFitResult* fr) :
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooHist.cxx]: A doxygen-readable comment would be nice, such as:
```
/// Constructor with ...
/// 
/// More details.
/// \param[in] f The variable to be plotted.
/// \param[in] x ...
```

**Comment by hageboeck** [comment]
We cannot put merge commits in master. The branch needs to be rebased interactively to drop the merge commit, and afterwards, it needs to be rebased on top of master.

File: roofit/roofitcore/src/RooPlot.cxx (hunk 3 above)
Comment (cburgard) [reply roofit/roofitcore/src/RooPlot.cxx]: This was mostly added to provide a consistent interface withe the other function defined in the next line.

File: roofit/roofitcore/src/RooPlot.cxx (hunk 4 above)
Comment (cburgard) [reply roofit/roofitcore/src/RooPlot.cxx]: Will be done.

File: roofit/roofitcore/src/RooHist.cxx (hunk 5 above)
Comment (cburgard) [reply roofit/roofitcore/src/RooHist.cxx]: Will be done.

**Comment by cburgard** [comment]
Documentation was added and the branch was rebased to master. Anything else I should do?

File: roofit/roofit/src/RooBernstein.cxx [hunk 6]
```diff
@@ -82,16 +83,32 @@ RooBernstein::RooBernstein(const char* name, const char* title,
...
 {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 
+
+void RooBernstein::selectNormalizationRange(const char* rangeName, Bool_t force)
+{
+  if (rangeName && (force || !_refRangeName)) {
+     _refRangeName = (TNamed*) RooNameReg::instance().constPtr(rangeName);
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx]: static_cast<TNamed*>(...)

File: roofit/roofit/inc/RooBernstein.h [hunk 7]
```diff
@@ -33,15 +33,17 @@ class RooBernstein : public RooAbsPdf {
 
   Int_t getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVars, const char* rangeName=0) const ;
   Double_t analyticalIntegral(Int_t code, const char* rangeName=0) const ;
+  virtual void selectNormalizationRange(const char* rangeName=0, Bool_t force=kFALSE) ;
 
 private:
 
   RooRealProxy _x;
   RooListProxy _coefList ;
+  mutable TNamed* _refRangeName ;
```
Comment (hageboeck) [inline roofit/roofit/inc/RooBernstein.h]: Please don't make it mutable.

File: roofit/roofit/src/RooBernstein.cxx [hunk 8]
```diff
@@ -137,10 +154,12 @@ Double_t RooBernstein::evaluate() const
 Int_t RooBernstein::getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVars, const char* rangeName) const
 {
   if (rangeName && strlen(rangeName)) {
-    return 0 ;
+  _refRangeName = 0;
+  return 0 ;
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx]: This looks wrong. Why should a call to the integral remove the range?

File: roofit/roofit/src/RooBernstein.cxx [hunk 9]
```diff
@@ -149,7 +168,10 @@ Int_t RooBernstein::getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVar
 Double_t RooBernstein::analyticalIntegral(Int_t code, const char* rangeName) const
 {
   R__ASSERT(code==1) ;
-  Double_t xmin = _x.min(rangeName); Double_t xmax = _x.max(rangeName);
+//  const Double_t xmax = _x.max(_refRangeName?_refRangeName->GetName():0);
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx]: This ignores ranges coming from the outside. Is that the intention?

File: roofit/roofit/src/RooBernstein.cxx [hunk 10]
```diff
@@ -51,7 +51,7 @@ ClassImp(RooBernstein);
 
 ////////////////////////////////////////////////////////////////////////////////
 
-RooBernstein::RooBernstein()
+RooBernstein::RooBernstein() : _refRangeName(0)
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx]: nullptr

File: roofit/roofit/src/RooBernstein.cxx [hunk 11]
```diff
@@ -62,7 +62,8 @@ RooBernstein::RooBernstein(const char* name, const char* title,
                            RooAbsReal& x, const RooArgList& coefList):
   RooAbsPdf(name, title),
   _x("x", "Dependent", this, x),
-  _coefList("coefficients","List of coefficients",this)
+  _coefList("coefficients","List of coefficients",this),
+  _refRangeName(0)
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx]: nullptr

File: roofit/roofitcore/src/RooHist.cxx [hunk 12]
```diff
@@ -260,6 +261,74 @@ RooHist::RooHist(const RooHist& hist1, const RooHist& hist2, Double_t wgt1, Doub
...
+  funcPtr= f.bindVars(x,normVars,kTRUE);
+
+  // apply a scale factor if necessary
+  if(scaleFactor != 1) {
+    rawPtr= funcPtr;
+    funcPtr= new RooScaledFunc(*rawPtr,scaleFactor);
+  }
+  
+  // apply a scale factor if necessary
+  assert(0 != funcPtr);
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooHist.cxx]: assert(funcPtr)

**Comment by cburgard** [comment]
> Hello Rahul,
> 
> I had a look over the changes. and added some comments.
> Next time, please don't mix different topics such as graphics and Bernstein ranges, as the continuous integration should test them one by one.

The "mixing of topics" is my fault here - this is our development branch for the RooFit implementation of RooUnfold, and so far we only needed to change graphics. We encountered problems with RooBernstein and needed to fix them to get our examples running. We hope that not so many further changes are needed, but we will try to keep topics more separate in the future. If you want, I can split this into two different branches before merging.

**Comment by hageboeck** [comment]
Hello all,

there has apparently been a rebase error, no need to review at the moment!

**Comment by hageboeck** [comment]
Hello @cburgard and Rahul,

there must have been an error when rebasing the branch, would you try again with `root-project/master`, and only apply the RooFit-related commits?
You might also want to try rebase --interactive to reorder/squash commits. After that, let's go through the comments, because I completely lost track if we addressed them or there's still something to be done.

Are you expecting more changes or are we close to "finishing" this line of development?

**Comment by cburgard** [comment]
closing this PR for now to open a new one for a rebased branch

**Pull request closed without merging**